protocol.md

```diff
@@ -726,6 +727,141 @@ Upload-Length: 11
 Upload-Concat: final;/files/a /files/b
 ```
 
+### Challenge
+
+With this extension, Clients can authenticate follow-up requests to a [Creation](#creation) or [Creation With Upload](#creation-with-upload) 
+`POST` request using a high-entropy cryptographic random shared secret - and challenges
+based thereon.
+
+If the Server supports this extension, it MUST add `challenge` to the `Tus-Extension` header.
+
+#### Headers
+
+##### Upload-Secret
+
+The `Upload-Secret` MUST be a high-entropy cryptographic random string, consisting
+of 48 to 256 characters from the [base64 alphabet](https://tools.ietf.org/html/rfc4648#section-4).
+That's 36 to 192 bytes encoded as base64.
+
+The `Upload-Secret` header MUST NOT be included with anything but `POST` requests
+that create a new upload resource. Servers receiving an `Upload-Secret` header with
+any other request MUST respond with a `400 Bad Request` status.
+
+##### Upload-Challenge
+
+If an `Upload-Secret` header was sent with the `POST` request that created
+the upload resource, Clients MUST include an `Upload-Challenge` header with every
+subsequent request targeting that upload resource.
+
+The value of `Upload-Challenge` is defined as follows:
+
+```
+Upload-Challenge = SHA256([Upload-Secret] + SHA256(`[HTTP Method]` + [Upload-Offset] + [Upload-Secret] + [Content-Length]))
+```
```

Comment:
On the same topic as above. SHA256 might be OK now but maybe not in the future. We should mimic how the Checksum extension works and have the server specify one (or several) checksum algorithms that is supported for this extension. The checksum extension adds a `Tus-Checksum-Algorithm` header to the OPTIONS request with a comma separated list of algorithms supported. Could we add the same thing here, `Tus-Challenge-Algorithm`?

Reply:
That's a very good point, too.

What I think we should definitely do then is prefix the `Upload-Challenge` with the used checksum algorithm, like so

```
Upload-Challenge: sha256:59ca949f7eed4c4a1f7e40da7c119fb7cb1095262a12eab770685711ea5a176a
```

An additional `Tus-Challenge-Algorithm` header could then provide a list of supported hashes in order of preference:

```
Tus-Challenge-Algorithm: sha256, sha1, md5
```

(I'd not really want `sha1` or `md5` supported here - using them just as examples)

However, considering that `sha256` would initially be the only supported algorithm and adding (and then requiring) new algorithms would require server and client code changes (i.e. be a "breaking" change), I see several choices:

1) include `Tus-Challenge-Algorithm` in the spec, but make it optional for clients and servers. Since adding a new algorithm would be a breaking change anyway, the additional header could also still be switched to being required at that later point. 
2) leave out `Tus-Challenge-Algorithm` from the spec to maximize simplicity and remain flexible, depending on what is needed at the time a new algorithm needs to be supported.

Personally, I'd prefer 1), but could also agree to 2).

Reply:
I think we could copy the checksum extension pretty much 100% here. It basically already solves this issue in a (IMO) nice and clean way:

> The Server MUST support at least the SHA1 checksum algorithm identified by sha1. The names of the checksum algorithms MUST only consist of ASCII characters with the modification that uppercase characters are excluded.
> 
> The Tus-Checksum-Algorithm header MUST be included in the response to an OPTIONS request.

So forcing the Tus-Challenge-Algorithm header and everyone must implement sha256 (but are allowed to add others) and then prefix the challenge with what algorithm that was used:

```
OPTIONS /files HTTP/1.1
Tus-Extensions: challenge
Tus-Challenge-Algorithm: sha256, sha512

PATCH /files/17f44dbe1c4bace0e18ab850cf2b3a83 HTTP/1.1
Content-Length: 11
Upload-Offset: 0
Tus-Resumable: 1.0.0
Upload-Challenge: sha256 <checksum goes here>
```

Reply:
I agree with a more flexible solution using `Tus-Challenge-Algorithm` while also requiring support for the SHA-256 hash. As @smatsson said, we could copy a lot from the Checksum extension.

Reply:
Sounds all good to me!